HACKING.adoc

```diff
@@ -302,6 +302,27 @@ If `boot/ocamlc` changes (e.g. because you ran `make bootstrap`), then
 the build will revert to the slower bytecode-compiled `ocamlc` until
 you do the above step again.
 
+=== Using merlin
+
+When building the compiler as described above, the cmi and cmt files that are
+produced are 99% of the time not readable by merlin, because it is usually
+updated to support new versions of the compiler after they get released.
+
+Therefore, to use merlin on the compiler, you want to build the compiler with an
+older version of itself. The easiest way to do this is to install the latest
+released version of the compiler, install dune, and run:
+
+----
+dune build @libs
+----
+
+which will do a bytecode build of all the distribution (without linking the
+executables), as well as generate a .merlin file.
+
+N.B. for this to work properly, you need to have run the configure script, and
+the workspace needs to be clean (otherwise dune will complain that build
+artefacts are present among the sources).
```

Comment:
Correct me if I'm wrong, but the `.merlin` files will lookup the build artefacts generated by Dune, and not at all those generated by Make. This means that if (I am developing with Make and) I change a module API and I want Merlin to see the change, I need to run `make clean` first, and then do the dune thing. This should be clarified/documented, because otherwise users (including developers) are going to get confused.

Is there a way to ask Dune to ignore build artefacts, so that we don't have to make clean (and lose make incremental builds) to get Merlin to notice an API change?

Reply:
> Correct me if I'm wrong, but the .merlin files will lookup the build artefacts generated by Dune, and not at all those generated by Make. This means that if (I am developing with Make and) I change a module API and I want Merlin to see the change, I need to run make clean first, and then do the dune thing.

Correct.

> This should be clarified/documented, because otherwise users (including developers) are going to get confused.

I will try.

> Is there a way to ask Dune to ignore build artefacts, so that we don't have to make clean (and lose make incremental builds) to get Merlin to notice an API change?

Not to my knowledge. And I think it might not be an easy one to add. Let's ping @diml, see what he says.